### PR TITLE
Database migration from classic mode

### LINKED ISSUE #2826: Database migration from classic mode
With 1.7.x classic mode only `URI_REGISTRY` table is used and `APP_REGISTRATION` stays empty as it's only used with skipper mode. As 2.x requires skipper, `URI_REGISTRY` table is not currently migrated.

Should try to tweak flyway migrations so that registered apps are in place with this combination as well.

### LINKED ISSUE #2863: Possible double insert into hibernate_sequence with mysql
Both dataflow and skipper does a simple `insert into hibernate_sequence values ( 1 )` which is causing two rows in a table as mysql don't have a proper sequence. Should make sure that double insert doesn't happen.

### LINKED ISSUE #2868: Migration clob fails with postgresql
For tables in 1.7.x like `STREAM_DEFINITIONS`, some field types are `text` as they are also in 2.x with JPA. However in 2.x entity class type for field is `@Lob` which causes wrapped CLOB and data between migrated tables cannot be copied as is.

Reason for this is because with postgres real data in a field is a long which is an oid to a large object.

### PR DESCRIPTION
**Pull request by jvalkeal**
- As mysql don't have a real sequence, initial value
  needs to be inserted. This change makes sure that
  we only insert value once and that value will be 1.
  Similar fix will also be done on a skipper side.
- Fix clob handling with postgres.
- New base impl AbstractMigrateUriRegistrySqlCommand which
  all db's implement to support migration from URI_REGISTRY
  to app_registration table.
- Fixes #2863
- Fixes #2826
- Fixes #2868

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/db/migration/AbstractMigrateUriRegistrySqlCommand.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+				if (!defaultVersions.contains(key)) {
+					armd.setDefaultVersion(true);
+					defaultVersions.add(key);
+				}
+				if (split.length == 2) {
+					// we got *.* first
+					armd.setName(split[1]);
+					armd.setType(ApplicationType.valueOf(split[0]).ordinal());
+					armd.setUri(uri);
+					armd.setVersion(arc.getResourceVersion(arc.getResource(uri)));
```
Comment (markpollack) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/db/migration/AbstractMigrateUriRegistrySqlCommand.java]: I guess we need to discuss how to inform users who are in classic mode and don't have a version field that their stream def isn't going to be migrated.  Also, I believe the old table gets dropped, so we prob shouldn't do that if we could not migrate a registered app....

**Comment by ilayaperumalg** [comment]
I went through the upgrade scenarios with MySQL, DB2, Postgres and MSSQL so far.

With MSSQL, I see some issues after the migration.

Originally I had SCDF classic mode server `1.7.4.RELEASE` with a stream definition:

```
dataflow:>stream list
╔═══════════╤═════════════════╤═════════════════════════════════════════╗
║Stream Name│Stream Definition│                 Status                  ║
╠═══════════╪═════════════════╪═════════════════════════════════════════╣
║a1         │time | log       │The stream has been successfully deployed║
╚═══════════╧═════════════════╧═════════════════════════════════════════╝

```

After the migration on SCDF 2.x, I see this:

```
dataflow:>stream list
Command failed org.springframework.cloud.dataflow.rest.client.DataFlowClientException: 115E:(pos 0): unexpected data in stream definition '楴'
楴敭簠氠杯
^
```

The `app list` shows the correct information after the migration though.

**Comment by jvalkeal** [comment]
Haven't seen that and just tried it manually. Maybe something to do with DB language settings running on your system.

**Comment by jvalkeal** [comment]
You'd need poke around database tables to see if there's something weird.

**Comment by ilayaperumalg** [comment]
The database table `stream_definitions` seems to have the correct entry:

```
definition_name                                                                                                                                                                                                                                                 definition
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
a1                                                                                                                                                                                                                                                              time | log

(1 rows affected)
```

**Comment by ilayaperumalg** [comment]
If the SCDF server is started before Skipper server (with SQLServer) then we see this:

```
Caused by: org.flywaydb.core.api.FlywayException: Error while executing beforeBaseline callback: StatementCallback; uncategorized SQLException for SQL [select next value for hibernate_sequence]; SQL state [S0001]; error code [11726]; Object 'hibernate_sequence' is not a sequence object.; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: Object 'hibernate_sequence' is not a sequence object.

```

**Comment by ilayaperumalg** [comment]
Created https://github.com/spring-cloud/spring-cloud-dataflow/issues/2903 for the hibernate_sequence table issue.

**Comment by ilayaperumalg** [comment]
MSSQL server issue mentioned above disappeared after using the correct SQL server docker image that has the necessary temp tables. 

LGTM, merging this PR.

**Comment by ilayaperumalg** [comment]
Rebased and merged as 9cdef0f7

**Pull request closed without merging**